photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.function
+
+import breeze.linalg.Vector
+import com.linkedin.photon.ml.data.{LabeledPoint, ObjectProvider}
+import com.linkedin.photon.ml.normalization.NormalizationContext
+import org.apache.spark.rdd.RDD
+
+/**
+  * The objective loss function for an optimization problem.
+  *
+  * @param data The dataset that the problem is optimized over
+  * @param normalizationContext The normalization context
+  * @param treeAggregateDepth The depth used by treeAggregate. Depth 1 indicates normal linear aggregate. Using
+  *                           depth > 1 can reduce memory consumption in the Driver and may also speed up the
+  *                           aggregation. It is experimental currently because treeAggregate is unstable in Spark
+  *                           versions 1.4 and 1.5.
+  */
+abstract class ObjectiveFunction(
+    protected[ml] val data: Either[RDD[LabeledPoint], Iterable[LabeledPoint]],
```

Comment:
We (Brendan, Yitong and I) once discussed a refactoring plan on ```DiffFunction``` to get rid of ```Either``` and use polymorphism instead in such scenario.

I should've discussed this with you earlier, because this ```DiffFunction``` redesign and refactoring was a good opportunity to make it happen. But let's leave ```Either``` as is in this PR and address this problem later.